### PR TITLE
Add named properties handling

### PR DESCRIPTION
**Pull request by RRozak**
It adds handling of named properties by replacing property references with its expressions. This approach was mentioned in IEEE Std 1800-2012, section 16.12.1.
Nested properties are supported, but not recursive.

### CODE REVIEW

**Comment by RRozak** [comment]
I will remove `test_regress/t/t_assert_named_property_simple`, because it is included in `t_assert_named_property`.

File: src/V3AssertPre.cpp [hunk 1]
```diff
@@ -67,6 +69,71 @@ class AssertPreVisitor final : public VNVisitor {
         m_senip = nullptr;
         m_disablep = nullptr;
     }
+    AstPropClocked* getPropertyExprp(const AstProperty* const propp) {
+        // The only statements possible in AstProperty are AstPropClocked (body)
+        // and AstVar (arguments).
+        AstNode* propExprp = propp->stmtsp();
+        while (VN_IS(propExprp, Var)) propExprp = propExprp->nextp();
```
Comment (kbieganski) [inline src/V3AssertPre.cpp:76]: Why not keep these in seperate `op` pointers? And just get them directly instead of iterating.

File: src/V3AssertPre.cpp [hunk 2]
```diff
@@ -88,10 +155,10 @@ class AssertPreVisitor final : public VNVisitor {
         iterateAndNextNull(nodep->stmtsp());
         m_seniAlwaysp = nullptr;
     }
-
```
Comment (kbieganski) [inline src/V3AssertPre.cpp:91]: Remove unnecessary whitespace changes.

File: src/V3AssertPre.cpp [hunk 3]
```diff
@@ -67,6 +69,71 @@ class AssertPreVisitor final : public VNVisitor {
...
+                // If disable iff is in outer property, move it to inner
+                if (nodep->disablep()) {
+                    AstNode* const disablep = nodep->disablep()->unlinkFrBack();
+                    propExprp->disablep(disablep);
+                }
+
+                if (nodep->sensesp() && propExprp->sensesp()) {
+                    nodep->v3warn(
+                        E_UNSUPPORTED,
+                        "Unsupported: Clock event before property call and in its instantion");
```
Comment (kbieganski) [inline src/V3AssertPre.cpp:121]:
```suggestion
                        "Unsupported: Clock event before property call and in its instantiation");
```

File: src/V3AssertPre.cpp [hunk 4]
```diff
@@ -189,6 +257,11 @@ class AssertPreVisitor final : public VNVisitor {
         // Reset defaults
         m_seniDefaultp = nullptr;
     }
+    void visit(AstProperty* nodep) override {
+        // The body will be visited when will be substituted in place of property reference
+        // (AstFuncRef)
+        // This node will be deleted in V3Assert (all references will be substituted then)
```
Comment (kbieganski) [inline src/V3AssertPre.cpp:263]: Note you can use `pushDeletep` to delete after the visitor is done.

File: src/V3AstNodeOther.h [hunk 5]
```diff
@@ -2427,6 +2427,14 @@ class AstFunc final : public AstNodeFTask {
...
     bool hasDType() const override { return true; }
 };
+class AstProperty final : public AstNodeFTask {
+    // A property inside a module
+public:
+    AstProperty(FileLine* fl, const string& name, AstNode* stmtp)
+        : ASTGEN_SUPER_Property(fl, name, stmtp) {}
+    ASTGEN_MEMBERS_AstProperty;
+    bool hasDType() const override { return true; }
+};
```
Comment (kbieganski) [inline src/V3AstNodeOther.h:2437]: Why do we inherit from `AstNodeFTask` if we don't use any task features? Also, could we just use `AstPropClocked`? Weird to have two node types with almost the same name, especially since this one doesn't do much new.

File: test_regress/t/t_assert_clock_event_bad.out [hunk 6]
```diff
@@ -0,0 +1,6 @@
+%Error-UNSUPPORTED: t/t_assert_clock_event_bad.v:26:7: Unsupported: Clock event before property call and in its instantion
+                                                     : ... In instance t
+   26 |       @(negedge clk)
+      |       ^
+                    ... For error description see https://verilator.org/warn/UNSUPPORTED?v=latest
+%Error: Exiting due to
```
Comment (kbieganski) [inline test_regress/t/t_assert_clock_event_bad.out:6]: I think this test should be called `*_unsup*` instead of `*_bad*`.

File: src/V3AssertPre.cpp [hunk 7]
```diff
@@ -67,6 +69,71 @@ class AssertPreVisitor final : public VNVisitor {
...
+                for (const auto& tconnect : tconnects) {
+                    const AstVar* const portp = tconnect.first;
+                    AstArg* const argp = tconnect.second;
+                    AstNode* const pinp = argp->exprp()->unlinkFrBack();
+                    replaceVarRefsWithExpr(propExprp, portp, pinp);
+                }
+                // Handle case with 2 disable iff statement (16.12.1)
+                if (nodep->disablep() && propExprp->disablep()) {
+                    nodep->v3error(
+                        "disable iff expression before property call and in its instantion");
```
Comment (kbieganski) [inline src/V3AssertPre.cpp:110]:
```suggestion
                        "disable iff expression before property call and in its instantiation not allowed");
```

File: src/V3Width.cpp [hunk 8]
```diff
@@ -4905,6 +4910,34 @@ class WidthVisitor final : public VNVisitor {
                                         // func
         }
     }
+    void visit(AstProperty* nodep) override {
+        if (nodep->didWidth()) return;
+        if (nodep->doingWidth()) {
+            UINFO(5, "Recursive property call: " << nodep);
+            nodep->v3warn(E_UNSUPPORTED,
+                          "Unsupported: Recursive property call: " << nodep->prettyNameQ());
```
Comment (kbieganski) [inline src/V3Width.cpp:4918]: Is this tested?

File: src/V3Width.cpp [hunk 9]
```diff
@@ -4905,6 +4910,34 @@ class WidthVisitor final : public VNVisitor {
...
+        // dtype, so set it here too
+        nodep->dtypeSetBit();
+        for (AstNode* propStmtp = nodep->stmtsp(); propStmtp; propStmtp = propStmtp->nextp()) {
+            if (VN_IS(propStmtp, Var))
+                userIterate(propStmtp, nullptr);
+            else if (VN_IS(propStmtp, PropClocked))
+                iterateCheckSizedSelf(nodep, "PropClocked", propStmtp, SELF, BOTH);
+            else
+                propStmtp->v3warn(E_UNSUPPORTED, "Unsupported: " << propStmtp->prettyNameQ()
+                                                                 << " in AstProperty");
```
Comment (kbieganski) [inline src/V3Width.cpp:4935]:
```suggestion
            if (VN_IS(propStmtp, Var)) {
                userIterate(propStmtp, nullptr);
            } else if (VN_IS(propStmtp, PropClocked)) {
                iterateCheckSizedSelf(nodep, "PropClocked", propStmtp, SELF, BOTH);
            } else {
                propStmtp->v3warn(E_UNSUPPORTED, "Unsupported: " << propStmtp->prettyNameQ()
                                                                 << " in AstProperty");
            }
```
Can the `else` actually be triggered by the user? If so, it should be tested, otherwise just do a `v3fatal("Invalid statement")` or something, as it should be unreachable.

**Review by kbieganski — CHANGES_REQUESTED** [review]

File: src/V3AssertPre.cpp (hunk 1 above)
Comment (RRozak) [reply src/V3AssertPre.cpp:76]: Because this class inherits from AstNodeFTask. I think it is needed, because Property references are parsed as AstFuncRef and linking rules are the same.

File: src/V3AssertPre.cpp (hunk 4 above)
Comment (RRozak) [reply src/V3AssertPre.cpp:263]: Ok, thanks.

File: test_regress/t/t_assert_clock_event_bad.out (hunk 6 above)
Comment (RRozak) [reply test_regress/t/t_assert_clock_event_bad.out]: I changed the name.

File: src/V3Width.cpp (hunk 9 above)
Comment (RRozak) [reply src/V3Width.cpp]: It should be unreachable. I changed the error.

File: src/V3Width.cpp (hunk 8 above)
Comment (RRozak) [reply src/V3Width.cpp:4918]: I added the test.

File: src/V3AstNodeOther.h (hunk 5 above)
Comment (RRozak) [reply src/V3AstNodeOther.h:2437]: Because Property references are parsed as AstFuncRef and linking rules are the same. Matching of arguments is also the same.
I think that I should find a better name for AstPropClocked. I think that we should have separate classes to distinguish between whole property (now AstProperty) and its expression (now AstPropClocked).

**Review by kbieganski — APPROVED** [review]
Please update the test outputs.

**Comment by RRozak** [comment]
Updated, thanks. I will close this issue and create new one when tests pass.

**Comment by RRozak** [comment]
Created PR to mainline: https://github.com/verilator/verilator/pull/3667.
